Improve defence against the external macros

davebayer opened the pull request:
This PR improves the resistence to the external macros coming from the Death Eaters and the Windows developers

Currently, we are trying to guard against public macros defined by Windows UCRT `max`, `min` and `interface` by wrapping the code with:
```cpp
_CCCL_PUSH_MACROS
// ...
_CCCL_POP_MACROS
```
Although it pushes the macros on the stack, it doesn't undefine them (as Eric pointed out).

Also, there are potential problems with collisions of attribute names and `__declspec` modifiers. In #4633, I've changed the pretty attribute names to their ugly form to prevent this, e. g.:
```cpp
// before #4633
#  if _CCCL_HAS_CPP_ATTRIBUTE(gnu::const)
#    define _CCCL_CONST [[gnu::const]]
// ...

// after #4633
#  if _CCCL_HAS_CPP_ATTRIBUTE(__gnu__::__const__)
#    define _CCCL_CONST [[__gnu__::__const__]]
// ...
```
Unfortunatelly `__declspec` doesn't offer any alternative to the pretty modifier names and if the user does:
```cpp
#define empty_bases
```
it will influence our code.

So, this PR removes `_CCCL_(PUSH|POP)_MACROS` macros and introduces `<cuda/std/__cccl/(push|pop)_macros.h>` includes instead that properly pushes the macros on the stack, makes sure those macros are undefined in the scope between those two includes and checks if a macro hasn't been reintroduced.

Also, from now on, we should guard all of the public code with `<cuda/std/__cccl/(push|pop)_macros.h>` includes. I've updated the formatting that should always put the includes in the right include group in the right order.

The internal header test now also tests for missing `<cuda/std/__cccl/pop_macros.h>` include.

![image](https://github.com/user-attachments/assets/8373366a-a17b-4dd0-97f8-b2ea52c1bbd2)

ericniebler commented:
one way to fix this would be to replace all occurrences of `_CCCL_PUSH_MACROS` with `#include <cuda/std/__cccl/prologue.hpp>` and similar for `_CCCL_POP_MACROS`. then we can put any code we want in the prologue and epilogue headers, including `#undef`s.

davebayer commented on libcudacxx/codegen/add_ptx_instruction.py:
```diff
@@ -90,12 +90,16 @@ def add_src(ptx_instr):
 #include <nv/target> // __CUDA_MINIMUM_ARCH__ and friends
 
+_CCCL_PUSH_MACROS
+
```
Oops, I didn't know that 👀

davebayer commented:
> one way to fix this would be to replace all occurrences of `_CCCL_PUSH_MACROS` with `#include <cuda/std/__cccl/prologue.hpp>` and similar for `_CCCL_POP_MACROS`. then we can put any code we want in the prologue and epilogue headers, including `#undef`s.

libc++ does
```cpp
_LIBCPP_PUSH_MACROS
#include <__undef_macros>

// ...

_LIBCPP_POP_MACROS
```

I would prefer this approach if you are fine with that :)

miscco commented:
/ok to test ace540f

davebayer commented:
pre-commit.ci autofix

davebayer commented:
pre-commit.ci autofix

gonidelis commented:
Any tracking issues or reference cases for further understanding of the problem solved?

davebayer commented:
> Any tracking issues or reference cases for the problem solved?

Sorry, I didn't realize I didn't provide any information and changed 750 files.. I've updated the description with the problem description and the implemented solution

miscco commented:
/ok to test 59eed24

miscco commented on libcudacxx/include/cuda/std/detail/libcxx/include/complex:
```diff
@@ -1594,6 +1594,8 @@ _CCCL_DIAG_POP
 
 _LIBCUDACXX_END_NAMESPACE_STD
 
+#include <cuda/std/__cccl/pop_macros.h>
+
 #if _LIBCUDACXX_HAS_NVFP16()
 #  include <cuda/std/__complex/nvfp16.h>
 #endif // _LIBCUDACXX_HAS_NVFP16()
```
I believe that the pop should come after the include

davebayer replied:
No, the includes have push/pop inside them

miscco commented:
/ok to test bb780df

ericniebler reviewed (changes requested):
:+1+ i use the same approach in `cudax/include/cuda/experimental/__execution`. these headers can be used for more than just pushing and popping macros. they are also a good place to disable warnings we don't care about, or define macros that we don't want leaking into user code.

i suggest renaming the headers to `prologue.h` and `epilogue.h`.

davebayer commented on libcudacxx/include/cuda/std/__cccl/push_macros.h:
```diff
@@ -0,0 +1,256 @@
+#  error \
+    "cccl internal error: <cuda/std/__cccl/pop_macros.h> must be included before next <cuda/std/__cccl/push_macros.h> is reincluded"
+#endif
+#define _CCCL_POP_MACROS_GUARD
```
It should be already handled by this check, because the first include defines the `_CCCL_POP_MACROS_GUARD` macro and the second include fails as it sees the `_CCCL_POP_MACROS_GUARD` macro defined which means there is an inclusion of `<pop_macros.h>` missing.

ericniebler commented:
/ok to test a0dfd7f

davebayer commented:
pre-commit.ci autofix

davebayer commented:
pre-commit.ci autofix

ericniebler commented:
/ok to test efa30b2

alliepiper commented on libcudacxx/include/cuda/__functional/maximum.h:
```diff
@@ -23,6 +23,8 @@
 
 #include <cuda/std/__type_traits/common_type.h>
 
+#include <cuda/std/__cccl/prologue.h>
```
Idea: This would definitely be for a follow up PR, but we should consider adding a test that scans libcu++ headers and verifies that these includes exist so we don't forget them in new headers. We have some machinery that does similar checks for Thrust/CUB: https://github.com/NVIDIA/cccl/blob/main/thrust/testing/cmake/check_source_files.cmake

miscco commented:
/ok to test 7dbbd5a

miscco commented:
pre-commit.ci autofix

oleksandr-pavlyk reviewed (approved):
Changes to the c/parallel sources/headers are purely reordering include statements, which is triggered by the change to `.clang-format` file.
